### PR TITLE
Add property-based tests for signers-voting contract functions

### PR DESCRIPTION
**Pull request by moodmosaic**
### Description

Add property-based tests for signers-voting contract functions

- reward-cycle-to-burn-height
- burn-height-to-reward-cycle
- is-in-prepare-phase

This effort, initially part of https://github.com/stacks-network/stacks-core/pull/4286, is now in a separate commit and PR as it's autonomous with regards to PoX-4 property and fuzz testing. Additional property-based tests for the remaining functions of the signers-voting contract are warranted and will be addressed in separate commits/PRs.

### Applicable issues

* https://github.com/stacks-network/stacks-core/pull/4286

### Checklist

- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

### CODE REVIEW

File: contrib/core-contract-tests/tests/pox-4/signers-voting.prop.test.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            "first-burnchain-block-height"];
+        const reward_cycle_length =
+          // @ts-ignore
+          pox_4_info.value.data[
+            "reward-cycle-length"];
+
+        // Act
+        const { result: actual } = simnet.callReadOnlyFn(
+          "signers-voting",
+          "reward-cycle-to-burn-height",
```
Comment (moodmosaic) [inline contrib/core-contract-tests/tests/pox-4/signers-voting.prop.test.ts:30]: Compared with the [existing tests in `signers-voting.test.ts`](https://github.com/stacks-network/stacks-core/blob/059ae88deae783081abdf159debcd2bd8d62e200/contrib/core-contract-tests/tests/pox-4/signers-voting.test.ts#L25-L35) file as well:

image

**Review by friedger — APPROVED** [review]
The tests for read-only functions look good.

Maybe variation over account is not necessary, as these are read-only functions that do not use tx-sender.

For a different PR, it might be interesting to check `is-in-prepare-phase` with block-height in a contract call vs read-only call.

**Review by hugocaillard — COMMENTED** [review]
I left a couple of suggestions.
Also, would you consider adding a formatter (such as prettier)?
Happy to help on that

File: contrib/core-contract-tests/tests/pox-4/signers-voting.prop.test.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      (account: string, reward_cycle: number) => {
+        // Arrange
+        const { result: pox_4_info } = simnet.callReadOnlyFn(
+          "pox-4",
+          "get-pox-info",
+          [],
+          account,
+        );
+        const first_burnchain_block_height =
+          // @ts-ignore
```
[the remark this answers is not in the document]
Comment (moodmosaic) [reply contrib/core-contract-tests/tests/pox-4/signers-voting.prop.test.ts:19]: Thanks! How do I get the actual value out though?

**Comment by moodmosaic** [comment]
> I left a couple of suggestions.

Thanks for the review!

>Also, would you consider adding a formatter (such as prettier)? Happy to help on that

I always run `deno ftm` and go with the defaults from that. Though if we have some existing [config for Prettier](https://prettier.io/docs/en/configuration.html), happy to use that. (Do we have one?)

**Comment by hugocaillard** [comment]
Any formater is good as far as i'm concerned. But I was a bit surprised with this formatting

```ts
          pox_4_info.value.data[
            "first-burnchain-block-height"];
```

**Comment by moodmosaic** [comment]
> Any formater is good as far as i'm concerned. But I was a bit surprised with this formatting
> 
> ```ts
>           pox_4_info.value.data[
>             "first-burnchain-block-height"];
> ```

me too, actually. It's left from when I was adding `// @ts-ignore` and making sure it won't ignore more than it should. Thanks 👍After I'm done checking https://github.com/stacks-network/stacks-core/pull/4403#discussion_r1504130148 I'll reformat and push.

File: contrib/core-contract-tests/tests/pox-4/signers-voting.prop.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (hugocaillard) [reply contrib/core-contract-tests/tests/pox-4/signers-voting.prop.test.ts:19]: Same as your a doing currently.
You might also need
```ts
assert(isClarityType(reward_cycle_length, ClarityType.UInt));
assert(isClarityType(first_burnchain_block_height ClarityType.UInt));
```

To safely access  `.value` of these two values

File: contrib/core-contract-tests/tests/pox-4/signers-voting.prop.test.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+          [Cl.uint(reward_cycle)],
+          account,
+        );
+
+        // Assert
+        const expected = Cl.uint(
+          (reward_cycle * Number(reward_cycle_length.value)) +
+            Number(first_burnchain_block_height.value),
+        );
+        expect(actual).toEqual(expected);
```
[the remark this answers is not in the document]
Comment (moodmosaic) [reply contrib/core-contract-tests/tests/pox-4/signers-voting.prop.test.ts]: ✅ Done

**Review by hugocaillard — APPROVED** [review]
LGTM

**Review by friedger — APPROVED** [review]
:shipit:

**Review by setzeus — APPROVED** [review]

**Pull request merged**